src/middleware/csrf/index.ts

```diff
@@ -0,0 +1,48 @@
+import type { Context } from '../../context'
+import { HTTPException } from '../../http-exception'
+import type { MiddlewareHandler } from '../../types'
+
+type IsAllowedOriginHandler = (origin: string, context: Context) => boolean
+interface CSRFOptions {
+  origin?: string | string[] | IsAllowedOriginHandler
+}
+
+const isSafeMethodRe = /^(GET|HEAD)$/
+const isRequestedByFormElementRe =
+  /^\b(application\/x-www-form-urlencoded|multipart\/form-data|text\/plain)\b/
+
+export const csrf = (options?: CSRFOptions): MiddlewareHandler => {
+  const handler: IsAllowedOriginHandler = ((optsOrigin) => {
+    if (!optsOrigin) {
+      return (origin, c) => origin === new URL(c.req.url).origin
+    } else if (typeof optsOrigin === 'string') {
+      return (origin) => origin === optsOrigin
+    } else if (typeof optsOrigin === 'function') {
+      return optsOrigin
+    } else {
+      return (origin) => optsOrigin.includes(origin)
+    }
+  })(options?.origin)
+  const isAllowedOrigin = (origin: string | undefined, c: Context) => {
+    if (origin === undefined) {
+      // denied always when origin header is not present
+      return false
+    }
+    return handler(origin, c)
+  }
+
+  return async function cors(c, next) {
+    if (
+      !isSafeMethodRe.test(c.req.method) &&
+      isRequestedByFormElementRe.test(c.req.header('content-type') || '') &&
+      !isAllowedOrigin(c.req.header('origin'), c)
+    ) {
+      const res = new Response('Forbidden', {
+        status: 403,
+      })
+      throw new HTTPException(401, { res })
```

Comment:
I think that `401` should be `403`.

P.S.

However, if it remains as `401`, the tests do not fail. This is because if `{ res }` is set, it can be used and the status number in the first argument of `HTTPException` is ignored. Therefore, there is no problem for real use cases with this code. But, for making it the same as other middleware, I think it should be changed to `403`.